test-data/unit/check-classes.test

```diff
@@ -1723,3 +1723,61 @@ def foo(c: Type[C], d: Type[D]) -> None:
 [out]
 main: note: In function "foo":
 main:7: error: Revealed type is 'builtins.list[Type[__main__.B]]'
+
+[case testTypeMatchesOverloadedFunctions]
+from typing import Type, overload, cast
+class User: pass
+@overload
+def f(a: object) -> None: pass
+@overload
+def f(a: int) -> None: pass
+f(User)
+f(cast(Type[User], User))
+[builtins fixtures/classmethod.py]
+[out]
+
+[case testTypeMatchesGeneralTypeInOverloadedFunctions]
+from typing import Type, overload, cast
+class User: pass
+@overload
+def f(a: type) -> None: pass
+@overload
+def f(a: int) -> None: pass
+f(User)
+f(cast(Type[User], User))
+[builtins fixtures/classmethod.py]
+[out]
+
+[case testGeneralTypeDoesNotMatchSpecificTypeInOverloadedFunctions]
+from typing import Type, overload, cast
+class User: pass
+@overload
+def f(a: Type[User]) -> None: pass
+@overload
+def f(a: int) -> None: pass
+f(User)
+f(type(User))
+[builtins fixtures/classmethod.py]
+[out]
+main:8: error: No overload variant of "f" matches argument types [builtins.type]
```

Comment:
This would also be a good candidate for `# E:`, except that I think this should *not* be an error.